### ISSUE TITLE
Extend the Roslyn API to help recognizing colliding members

### ISSUE THREAD

**Issue by tvsonar** [dotnet/roslyn#9042]
We're in the process of writing an analyzer to check colliding interface members. See [spec](https://jira.sonarsource.com/browse/RSPEC-3444), or [PR](https://github.com/SonarSource-VisualStudio/sonarlint-roslyn-analyzer/pull/794). Considering the below code fragment, the analyzer reports an issue on the `IDerived` interface saying that you won't be able to use `SomeProperty` directly through this interface. (CS0229 is reported on the call site, but we can prevent this issue by properly defining the interfaces.)
```C#
public interface IBase1
{
  string SomeProperty { get; set; }
}
public interface IBase2
{
  string SomeProperty { get; set; }
}
public interface IDerived : IBase1, IBase2 // we are reporting here on the interface declaration
{
}
...
public void M(IDerived element)
{
  element.SomeProperty = ""; // compiler error CS0229, can't decide between IBase1.SomeProperty and IBase2.SomeProperty
}
```
To overcome this issue there are multiple possibilities
* always cast down the `IDerived` instance to one of the base interfaces,
* add a `SomeProperty` member to `IDerived` too, or 
* change one of the members.

To perform the required analysis, one needs to compare member signatures in the two base interfaces. This comparison is not trivial in case of generic members, because for instance in the below code fragment `T1` and `T2` are different type parameter symbols, thus, a mechanism is needed to substitute both of them with the same symbol before comparing them.
```C#
public interface IBase1<TX>
{
    string MethodGen<T1>(T1[,] parameter, TX[] arr); 
    void M();
}
public interface IBase2<TY>
{
    int MethodGen<T2>(T2[,] parameter, TY[] arr); 
}
public interface IDerived<T> : IBase1<T>, IBase2<T> // We should report on this interface
{
}
```

This comparison mechanism is already implemented in Roslyn in the [`MemberSignatureComparer`](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.CSharp/Symbols/MemberSignatureComparer.cs,c6139a6a0229f1b0), which internally uses [`TypeMap`](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.CSharp/Symbols/TypeMap.cs,02d8298ab4cdbb35) to perform the type substitution. I believe that this functionality (similarly to other useful pieces of the compiler, such as https://github.com/dotnet/roslyn/issues/6670) should be made public to facilitate the job of tool creators. There is no point in duplicating compiler logic in analyzers, all API should be available to ease the creation of valuable analyzers and improve the capabilities of VS.

BTW, the Roslyn source code has the described design issue in [`INamedTypeDefinition`](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis/PEWriter/Types.cs,666c0bc28d70e5f2), where both base interfaces have property `GenericParameterCount`.

**Comment by CyrusNajmabadi** [comment]
Needs an actual API proposal.
